File: apps/web/lib/issue-feedback.ts
import "server-only";

import { randomUUID } from "node:crypto";
import { withAuditedProjectWrite } from "@/lib/audit";
import { requireCurrentProjectPermission } from "@/lib/data";
import { correlationId } from "@/lib/observability";
import { publishProjectEvent } from "@/lib/project-events";
import { issueFeedbackInputSchema,planIssueFeedback } from "@/lib/issue-feedback-core";

export async function recordIssueFeedback(projectId: number,issueId: number,rawInput: unknown,requestId?: string|null) {
  const input = issueFeedbackInputSchema.parse(rawInput);
  const { user,access } = await requireCurrentProjectPermission(projectId,"issue:handle");
  return withAuditedProjectWrite({ projectId,teamId: access.teamId,requestId: correlationId(requestId),
    idempotencyKey: input.clientKey,actorUserId: user.id,action: "issue.feedback.record",resourceType: "issue",resourceId: String(issueId),
    input: { action: input.action,detectionId: input.detectionId,expectedVersion: input.expectedVersion },
    policyResult: { permission: "issue:handle" }
  }, async (client) => {
    const replay = (await client.query<Record<string,unknown>>(`select id,action,created_at as "createdAt" from issue_feedback
      where project_id=$1 and issue_id=$2 and client_key=$3`,[projectId,issueId,input.clientKey])).rows[0];
    if (replay) return { ...replay,replayed: true };
    const evidence = (await client.query<{
      issueProjectId: number; issueVersion: number; detectionId: number; originalLabel: string;
      algorithmDefinitionVersionId: number; taskVersionId: number|null; taskRunStepId: number|null;
    }>(`select issue.project_id as "issueProjectId",issue.state_version as "issueVersion",detection.id::int as "detectionId",
      detection.label as "originalLabel",run.algorithm_definition_version_id::int as "algorithmDefinitionVersionId",
      issue.task_version_id::int as "taskVersionId",(
        select case when link.target_id~'^[0-9]+$' then link.target_id::bigint end from issue_links link
        where link.project_id=issue.project_id and link.issue_id=issue.id and link.link_type='task_step' order by link.id desc limit 1
      )::int as "taskRunStepId"
      from issues issue join issue_links detection_link on detection_link.project_id=issue.project_id
        and detection_link.issue_id=issue.id and detection_link.link_type='detection'
      join detections detection on detection.project_id=detection_link.project_id
        and detection.id=case when detection_link.target_id~'^[0-9]+$' then detection_link.target_id::bigint end
      join algorithm_runs run on run.id=detection.algorithm_run_id and run.project_id=detection.project_id
      where issue.project_id=$1 and issue.id=$2 and detection.id=$3 for update of issue`,[projectId,issueId,input.detectionId])).rows[0];
    if (!evidence) throw new Error("ISSUE_FEEDBACK_EVIDENCE_NOT_FOUND");
    const plan = planIssueFeedback(input,{ ...evidence,projectId });
    const feedback = (await client.query<{ id: number; createdAt: string }>(`insert into issue_feedback(
      project_id,team_id,issue_id,detection_id,algorithm_definition_version_id,task_version_id,task_run_step_id,
      action,corrected_label,disposition,reason,client_key,evidence_snapshot_json,actor_user_id)
      values($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14) returning id,created_at as "createdAt"`,
      [projectId,access.teamId,issueId,input.detectionId,evidence.algorithmDefinitionVersionId,evidence.taskVersionId,
        evidence.taskRunStepId,input.action,input.correctedLabel ?? null,input.disposition ?? null,input.reason,input.clientKey,
        plan.evidenceSnapshot,user.id])).rows[0];
    const eventType = `feedback.${input.action}`;
    await client.query(`insert into issue_events(project_id,issue_id,event_type,body,metadata_json,actor_user_id,client_key)
      values($1,$2,$3,$4,$5,$6,$7)`,[projectId,issueId,eventType,input.reason,{ feedbackId: feedback.id,...plan.evidenceSnapshot,
        correctedLabel: input.correctedLabel ?? null,disposition: input.disposition ?? null },user.id,`feedback:${input.clientKey}`]);
    await client.query(`update issues set state_version=state_version+1,updated_at=now() where project_id=$1 and id=$2`,[projectId,issueId]);
    await publishProjectEvent(client,{ projectId,teamId: access.teamId,eventId: randomUUID(),eventType: "issue.updated",
      payload: { issueId,feedbackId: feedback.id,feedbackAction: input.action,stateVersion: plan.issuePatch.stateVersion },enqueue: false });
    return { id: feedback.id,action: input.action,createdAt: feedback.createdAt,stateVersion: plan.issuePatch.stateVersion,replayed: false };
  });
}

File: apps/server/internal/httpapi/issue_feedback.go
package httpapi

import (
	"aerosight/server/internal/database"
	"aerosight/server/internal/database/sqlcgen"
	"database/sql"
	_ "embed"
	"encoding/json"
	"errors"
	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"io"
	"strconv"
	"strings"
)

//go:embed issue_feedback_input.json
var issueFeedbackSchema []byte

func (s *Server) issueFeedback(c *gin.Context) {
	fail := func(e error) {
		code := e.Error()
		status := 400
		if code == "ISSUE_VERSION_CONFLICT" {
			status = 409
		} else if strings.Contains(code, "ACCESS") || strings.Contains(code, "PERMISSION") {
			status = 403
		} else if strings.Contains(code, "NOT_FOUND") {
			status = 404
		}
		if !strings.HasPrefix(code, "ISSUE_") && code != "PROJECT_ACCESS_DENIED" {
			code = "ISSUE_FEEDBACK_FAILED"
		}
		s.failure(c, status, code)
	}
	pid, e := projectID(c)
	if e != nil {
		fail(errFHInput)
		return
	}
	iid, e := strconv.ParseInt(c.Param("issueId"), 10, 32)
	if e != nil || iid <= 0 {
		fail(errFHInput)
		return
	}
	raw, e := io.ReadAll(c.Request.Body)
	if e != nil {
		fail(errFHInput)
		return
	}
	input, e := parseFHInput(raw, issueFeedbackSchema)
	if e != nil {
		fail(errFHInput)
		return
	}
	action := fhString(input["action"])
	if action == "category_correction" && input["correctedLabel"] == nil || action != "category_correction" && input["correctedLabel"] != nil || action == "disposition" && input["disposition"] == nil {
		fail(errors.New("ISSUE_FEEDBACK_INPUT_INVALID"))
		return
	}
	key, e := uuid.Parse(fhString(input["clientKey"]))
	if e != nil {
		fail(errFHInput)
		return
	}
	ctx, uid := c.Request.Context(), currentUser(c).ID
	a, e := s.projectAccess(ctx, s.queries, uid, pid, "issue:handle")
	if e != nil {
		fail(e)
		return
	}
	did := fhOptionalNumber(input, "detectionId")
	audit := database.AuditContext{ProjectID: pid, TeamID: a.TeamID, ActorUserID: uid, RequestID: c.GetHeader("X-Request-ID"), IdempotencyKey: key.String(), Action: "issue.feedback.record", ResourceType: "issue", ResourceID: strconv.FormatInt(iid, 10), Input: gin.H{"action": action, "detectionId": did, "expectedVersion": input["expectedVersion"]}, PolicyResult: map[string]any{"permission": "issue:handle"}}
	result, e := database.AuditedWrite(ctx, s.db, audit, s.authorizeWrite(uid, pid, a.TeamID, "issue:handle", false), func(w *database.WriteTx) (gin.H, error) {
		q := w.Queries
		raw, e := q.FeedbackReplay(ctx, sqlcgen.FeedbackReplayParams{P1: pid, P2: int32(iid), P3: key})
		if e != nil {
			return nil, e
		}
		if len(raw) > 0 {
			r, e := fhFirstRow(raw, nil, "")
			if e != nil {
				return nil, e
			}
			r["replayed"] = true
			return r, nil
		}
		raw, e = q.FeedbackEvidence(ctx, sqlcgen.FeedbackEvidenceParams{P1: pid, P2: int32(iid), P3: did})
		row, e := fhFirstRow(raw, e, "ISSUE_FEEDBACK_EVIDENCE_NOT_FOUND")
		if e != nil {
			return nil, e
		}
		if row["issueProjectId"] != float64(pid) || row["detectionId"] != float64(did) {
			return nil, errors.New("ISSUE_FEEDBACK_SCOPE_MISMATCH")
		}
		if row["issueVersion"] != input["expectedVersion"] {
			return nil, errors.New("ISSUE_VERSION_CONFLICT")
		}
		evidence := gin.H{}
		for _, k := range []string{"detectionId", "originalLabel", "algorithmDefinitionVersionId", "taskVersionId", "taskRunStepId"} {
			evidence[k] = row[k]
		}
		evJSON, _ := json.Marshal(evidence)
		feedback, e := q.FeedbackInsert(ctx, sqlcgen.FeedbackInsertParams{P1: pid, P2: a.TeamID, P3: int32(iid), P4: did, P5: fhOptionalNumber(row, "algorithmDefinitionVersionId"), P6: sql.NullInt64{Int64: fhOptionalNumber(row, "taskVersionId"), Valid: row["taskVersionId"] != nil}, P7: sql.NullInt64{Int64: fhOptionalNumber(row, "taskRunStepId"), Valid: row["taskRunStepId"] != nil}, P8: action, P9: sql.NullString{String: fhString(input["correctedLabel"]), Valid: input["correctedLabel"] != nil}, P10: sql.NullString{String: fhString(input["disposition"]), Valid: input["disposition"] != nil}, P11: fhString(input["reason"]), P12: key, P13: evJSON, P14: uid})
		if e != nil {
			return nil, e
		}
		evidence["feedbackId"], evidence["correctedLabel"], evidence["disposition"] = feedback.ID, input["correctedLabel"], input["disposition"]
		evJSON, _ = json.Marshal(evidence)
		if e = q.FeedbackEvent(ctx, sqlcgen.FeedbackEventParams{P1: pid, P2: int32(iid), P3: "feedback." + action, P4: sql.NullString{String: fhString(input["reason"]), Valid: true}, P5: evJSON, P6: sql.NullInt32{Int32: uid, Valid: true}, P7: sql.NullString{String: "feedback:" + key.String(), Valid: true}}); e != nil {
			return nil, e
		}
		if e = q.FeedbackUpdate(ctx, sqlcgen.FeedbackUpdateParams{P1: pid, P2: int32(iid)}); e != nil {
			return nil, e
		}
		version := fhOptionalNumber(input, "expectedVersion") + 1
		_, e = w.Publish(ctx, database.ProjectEvent{ProjectID: pid, TeamID: a.TeamID, EventID: uuid.NewString(), EventType: "issue.updated", Payload: gin.H{"issueId": iid, "feedbackId": feedback.ID, "feedbackAction": action, "stateVersion": version}, NoEnqueue: true})
		return gin.H{"id": feedback.ID, "action": action, "createdAt": timestamp(feedback.CreatedAt), "stateVersion": version, "replayed": false}, e
	})
	if e != nil {
		fail(e)
		return
	}
	c.JSON(200, result)
}
